store/postgres/migrations/2024-01-05-170000_ds_corruption_fix_up/up.sql

```diff
@@ -0,0 +1,96 @@
+  JOIN deployment_schemas AS d ON sd.deployment = d.subgraph
+ WHERE d.name IN (SELECT relnamespace::regnamespace::name
+                    FROM pg_class
+                   WHERE relname LIKE 'data_sources$%' AND relkind = 'r'
+                 );
+
+-- check if the block numbers in the tables are OK
+-- if not, it can be
+-- - either the lower bound is higher than the last block
+-- - or the lower is fine, but the higher is past the last block
+DO $do$
+DECLARE
+    s text;
+BEGIN
+    FOR s IN SELECT deployment_schema FROM temp_sgd_last_block
+    LOOP
+        EXECUTE format($$
+        UPDATE temp_sgd_last_block
+           SET is_ok = NOT EXISTS (SELECT 1
+                                     FROM %1$I."data_sources$"
+                                    WHERE lower(block_range) > last_block_from_registry
+                                    LIMIT 1)
+                       AND NOT EXISTS (SELECT 1
+                                         FROM %1$I."data_sources$"
+                                        WHERE upper(block_range) > last_block_from_registry 
+                                          AND lower(block_range) <= last_block_from_registry
+                                        LIMIT 1),
+               is_upper_limit_too_high = EXISTS (SELECT 1
+                                                   FROM %1$I."data_sources$"
+                                                  WHERE upper(block_range) > last_block_from_registry 
+                                                    AND lower(block_range) <= last_block_from_registry
+                                                  LIMIT 1)
+         WHERE deployment_schema = '%1$s'  
+                $$, s);
+    END LOOP;
+END;
+$do$;
+
+SELECT * FROM temp_sgd_last_block WHERE NOT is_ok;
+
+DO $do$
+DECLARE
+    schema text;
+    last_block_from_registry integer;
+    cnt bigint;
+BEGIN
+    FOR schema, last_block_from_registry IN SELECT deployment_schema, t.last_block_from_registry
+                                           FROM temp_sgd_last_block AS t
+                                          WHERE NOT is_ok AND NOT is_upper_limit_too_high
+    LOOP
+        EXECUTE format($$SELECT count(9) FROM %I."data_sources$"$$, schema) INTO cnt;
+        RAISE NOTICE 'before DELETE % has % rows', schema, cnt;
+        EXECUTE format($$DELETE FROM %I."data_sources$" WHERE lower(block_range) > %s $$, schema, last_block_from_registry);
+        GET DIAGNOSTICS cnt = ROW_COUNT;
+        RAISE NOTICE 'DELETEd % rows from %', cnt, schema;
+    END LOOP;
+
+    FOR schema, last_block_from_registry IN SELECT deployment_schema, t.last_block_from_registry 
+                                           FROM temp_sgd_last_block AS t
+                                          WHERE NOT is_ok AND is_upper_limit_too_high
+    LOOP
+        EXECUTE format($$UPDATE %I."data_sources$" 
+                            SET block_range = range_merge(block_range, (%2$s,))
```

Comment:
I _think_ this is setting the `block_range` to `int4range(lower(block_range), null)`, right? Why not just write that here?

Reply:
One can do that as well, indeed.  This must have been some atavism on my side, coming from Python coding, I guess :)